oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/web/ServletBearerExchangeFilterFunction.java

```diff
@@ -0,0 +1,247 @@
+				.switchIfEmpty(Mono.defer(() -> next.exchange(request)));
+	}
+
+	private Mono<ClientRequest> mergeRequestAttributesIfNecessary(ClientRequest request) {
+		if (request.attribute(AUTHENTICATION_ATTR_NAME).isPresent()) {
+			return Mono.just(request);
+		}
+
+		return mergeRequestAttributesFromContext(request);
+	}
+
+	private Mono<ClientRequest> mergeRequestAttributesFromContext(ClientRequest request) {
+		ClientRequest.Builder builder = ClientRequest.from(request);
+		return Mono.subscriberContext()
+				.map(ctx -> builder.attributes(attrs -> populateRequestAttributes(attrs, ctx)))
+				.map(ClientRequest.Builder::build);
+	}
+
+	private void populateRequestAttributes(Map<String, Object> attrs, Context ctx) {
+		RequestContextDataHolder holder = RequestContextSubscriber.getRequestContext(ctx);
+		if (holder == null) {
+			return;
+		}
+		if (holder.getAuthentication() != null) {
+			attrs.putIfAbsent(AUTHENTICATION_ATTR_NAME, holder.getAuthentication());
+		}
+	}
+
+	private AbstractOAuth2Token getOAuth2Token(Map<String, Object> attrs) {
+		Authentication authentication = (Authentication) attrs.get(AUTHENTICATION_ATTR_NAME);
+		if (authentication.getCredentials() instanceof AbstractOAuth2Token) {
+			return (AbstractOAuth2Token) authentication.getCredentials();
+		}
+		return null;
+	}
+
+	private ClientRequest bearer(ClientRequest request, AbstractOAuth2Token token) {
+		return ClientRequest.from(request)
+				.headers(headers -> headers.setBearerAuth(token.getTokenValue()))
+				.build();
+	}
+
+	private <T> CoreSubscriber<T> createRequestContextSubscriber(CoreSubscriber<T> delegate) {
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		return new RequestContextSubscriber<>(delegate, authentication);
+	}
+
+	private void populateDefaultAuthentication(Map<String, Object> attrs) {
+		if (attrs.containsKey(AUTHENTICATION_ATTR_NAME)) {
+			return;
+		}
+		Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+		attrs.putIfAbsent(AUTHENTICATION_ATTR_NAME, authentication);
+	}
+
+	private static class RequestContextDataHolder {
+		private final Authentication authentication;
+
+		RequestContextDataHolder(Authentication authentication) {
+			this.authentication = authentication;
+		}
+
+		public Authentication getAuthentication() {
+			return this.authentication;
+		}
+	}
+
+	private static class RequestContextSubscriber<T> implements CoreSubscriber<T> {
+		private static final String REQUEST_CONTEXT_DATA_HOLDER_ATTR_NAME =
+				RequestContextSubscriber.class.getName().concat(".REQUEST_CONTEXT_DATA_HOLDER");
+
+		private CoreSubscriber<T> delegate;
+		private final Context context;
+
+		private RequestContextSubscriber(CoreSubscriber<T> delegate,
+				Authentication authentication) {
+
+			this.delegate = delegate;
+			Context parentContext = this.delegate.currentContext();
+			Context context;
+			if (parentContext.hasKey(REQUEST_CONTEXT_DATA_HOLDER_ATTR_NAME)) {
```

Comment:
maybe it would be better not to create empty RequestContextDataHolder?
by 
```java
if (authentication == null || parentContext.hasKey(REQUEST_CONTEXT_DATA_HOLDER_ATTR_NAME)) {}
```